### ISSUE TITLE
"Wrong bind request from..." message on Hazelcast startup in Google's Kubernetes

### ISSUE THREAD

**Issue by barnyrelph** [hazelcast/hazelcast-kubernetes#31]
First, thanks so much for writing and maintaining this project, it's exactly what I need.

So I've written a simple Java app that uses a Hazelcast Lock to ensure that a task only runs on one instance (the others bale when the lock is held). The app sits in a permanent loop attempting to run the task every 2 seconds and the task itself takes 5 seconds, so I can see plenty of lock contention. 

This works fine on my local computer so I've now packaged it up as an image and am attempting to run it on Google's Container Engine. 

I'm configuring Hazelcast through Spring like so:

```
<hz:hazelcast id="instance">
		<hz:config>
			<hz:properties>
				<hz:property name="hazelcast.discovery.enabled">true</hz:property>
			</hz:properties>
			<hz:network port="5701" port-auto-increment="true">
				<hz:join>
					<hz:multicast enabled="false" />
					<hz:tcp-ip enabled="false" />

					<hz:discovery-strategies>

						<hz:discovery-strategy
							class-name="com.noctarius.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategy">
							<hz:properties>
								<hz:property name="service-dns">hazeltest-service-simple</hz:property>
								<hz:property name="service-dns-timeout">10</hz:property>
							</hz:properties>
						</hz:discovery-strategy>
					</hz:discovery-strategies>

				</hz:join>

			</hz:network>
		</hz:config>
	</hz:hazelcast>
```

The service definition I'm using looks like this:
```
apiVersion: v1
kind: Service
metadata:
  labels:
    name: hazeltest-name-label-simple
  name: hazeltest-service-simple
spec:
  ports:
    - port: 5701
      targetPort: 5701
      protocol: TCP
  selector:
    app: hazeltest-app-label
```

When I run the app, I can see the following in the logs:
```
2017-02-08T11:37:42.435702737Z Feb 08, 2017 11:37:42 AM com.hazelcast.nio.tcp.SocketAcceptorThread
2017-02-08T11:37:42.435758276Z INFO: [10.0.0.17]:5701 [dev] [3.7.5] Accepting socket connection from /10.0.1.12:52664
2017-02-08T11:37:42.461609384Z Feb 08, 2017 11:37:42 AM com.hazelcast.nio.tcp.TcpIpConnectionManager
2017-02-08T11:37:42.461679256Z INFO: [10.0.0.17]:5701 [dev] [3.7.5] Established socket connection between /10.0.0.17:5701 and /10.0.1.12:52664
2017-02-08T11:37:42.462780759Z Feb 08, 2017 11:37:42 AM com.hazelcast.nio.tcp.TcpIpConnectionManager
2017-02-08T11:37:42.462796237Z WARNING: [10.0.0.17]:5701 [dev] [3.7.5] Wrong bind request from [10.0.1.12]:5701! This node is not requested endpoint: [10.3.250.127]:5701
2017-02-08T11:37:42.464388597Z Feb 08, 2017 11:37:42 AM com.hazelcast.nio.tcp.TcpIpConnection
2017-02-08T11:37:42.464411912Z INFO: [10.0.0.17]:5701 [dev] [3.7.5] Connection[id=8, /10.0.0.17:5701->/10.0.1.12:52664, endpoint=null, alive=false, type=MEMBER] closed. Reason: Wrong bind request from [10.0.1.12]:5701! This node is not requested endpoint: [10.3.250.127]:5701
```

I'm running Hazelcast 3.7.5 and version 0.9.3 of this plugin, if that helps. Any pointers on why this would be happening or how to resolve would be super-appreciated.

**Comment by harels** [comment]
Try to configure using the kubernetes service API, worked for me:
```
<property name="service-name">hazeltest-service-simple</property>
<property name="namespace">default</property>
```

**Comment by noctarius** [comment]
Yeah it seems like GCE has different result when discovery is used against the DNS server instead of the kubernetes REST API.

**Comment by barnyrelph** [comment]
Nice one, thanks both! I'll give that a go now. I did try previously and got an SSL error, but I'll re-enable and  see what I get this time.

**Comment by amoAHCP** [comment]
@barnyrelph Did you turn on DNS for Kubernetes in GCloud? As far as I remember I had to do so.

**Comment by barnyrelph** [comment]
@amoAHCP Thanks for the suggestion. It looks like that's enabled by default in google's K8s currently, I can see kube-dns pods running on the cluster and I'm able to `nslookup hazeltest-service-simple` from within one of the pods.

**Comment by amoAHCP** [comment]
Try : <discovery-strategies>
                <discovery-strategy enabled="true"
                                    class="com.noctarius.hazelcast.kubernetes.HazelcastKubernetesDiscoveryStrategy">
                    <properties>
                        
                        <property name="service-dns">cluster.local</property>
                    </properties>
                </discovery-strategy>
            </discovery-strategies>
This should find all Endpoints deployed in your namespace.

**Comment by barnyrelph** [comment]
I've just tried to switch over to using the REST API properties. If I run it locally I get an exception because the token can't be found, which is fine. When I run the image in Google's K8s, I get the following exception:

```
2017-02-08T13:19:42.192510049Z Caused by: javax.net.ssl.SSLException: Received fatal alert: protocol_version
2017-02-08T13:19:42.192513976Z 	at sun.security.ssl.Alerts.getSSLException(Alerts.java:208)
2017-02-08T13:19:42.192517704Z 	at sun.security.ssl.Alerts.getSSLException(Alerts.java:154)
2017-02-08T13:19:42.192521665Z 	at sun.security.ssl.SSLSocketImpl.recvAlert(SSLSocketImpl.java:1989)
2017-02-08T13:19:42.192525627Z 	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1096)
2017-02-08T13:19:42.192529399Z 	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1342)
2017-02-08T13:19:42.192533070Z 	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1369)
2017-02-08T13:19:42.192536792Z 	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1353)
2017-02-08T13:19:42.192540437Z 	at com.squareup.okhttp.internal.io.RealConnection.connectTls(RealConnection.java:188)
2017-02-08T13:19:42.192544138Z 	at com.squareup.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:145)
2017-02-08T13:19:42.192547904Z 	at com.squareup.okhttp.internal.io.RealConnection.connect(RealConnection.java:108)
2017-02-08T13:19:42.192551611Z 	at com.squareup.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:174)
2017-02-08T13:19:42.192555470Z 	at com.squareup.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:126)
2017-02-08T13:19:42.192559347Z 	at com.squareup.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:95)
2017-02-08T13:19:42.192563065Z 	at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:283)
2017-02-08T13:19:42.192566745Z 	at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:224)
2017-02-08T13:19:42.192570415Z 	at com.squareup.okhttp.Call.getResponse(Call.java:286)
2017-02-08T13:19:42.192578552Z 	at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:243)
2017-02-08T13:19:42.192582801Z 	at io.fabric8.kubernetes.client.utils.HttpClientUtils$3.intercept(HttpClientUtils.java:110)
2017-02-08T13:19:42.192586594Z 	at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:232)
2017-02-08T13:19:42.192590264Z 	at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:205)
2017-02-08T13:19:42.192594009Z 	at com.squareup.okhttp.Call.execute(Call.java:80)
2017-02-08T13:19:42.192597672Z 	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:203)
```

Looking at that stack trace, it's coming from the fabric8 dependency (currently version 1.3.66). Looking at their issues, this got reported and apparently fixed in 1.3.102 : https://github.com/fabric8io/kubernetes-client/issues/472 

If I get a chance this afternoon I'll try to grab this repo and do a local build with the latest 1.3.x fabric8 release. Thanks all for your help on this!

**Comment by barnyrelph** [comment]
@amoAHCP is cluster.local a bind setting for hazelcast? sorry, I'm very new to Hazelcast and reasonably new to Kubernetes

**Comment by amoAHCP** [comment]
No, it has nothing to do with Hazelcast. cluster.local is a DNS entry (sorry I’m not a network specialist) returning all values registered currently. The Implementation of the Hazelcast plugin iterates over all entries returned by the lookup and tries to connect.

 

Von: Barnaby relph <notifications@github.com>
Antworten an: hazelcast/hazelcast-kubernetes <reply@reply.github.com>
Datum: Mittwoch, 8. Februar 2017 um 14:32
An: hazelcast/hazelcast-kubernetes <hazelcast-kubernetes@noreply.github.com>
Cc: Andy Moncsek <amo@mailbox.org>, Mention <mention@noreply.github.com>
Betreff: Re: [hazelcast/hazelcast-kubernetes] "Wrong bind request from..."message on Hazelcast startup in Google's Kubernetes (#31)

 

@amoAHCP is cluster.local a bind setting for hazelcast? sorry, I'm very new to Hazelcast and reasonably new to Kubernetes

**Comment by barnyrelph** [comment]
So I cloned this repo and changed the dependency on io.fabric8:kubernetes-client from 1.3.66 to 1.3.104 and re-built the project. I then updated my little project to use the locally built version of hazelcast-kubernetes. 

Reviewing the logs from the pods up on Google's K8s gets me the following:

```
2017-02-08T15:51:09.095363436Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Kubernetes Discovery properties: { service-dns: null, service-dns-timeout: 5, service-name: hazeltest-service-simple, service-label: name, service-label-value: hazeltest-name-label-simple, namespace: hazel-ns, kubernetes-master: https://kubernetes.default.svc}
2017-02-08T15:51:09.097827001Z Feb 08, 2017 3:51:09 PM com.hazelcast.spi.discovery.integration.DiscoveryService
2017-02-08T15:51:09.097842809Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Kubernetes Discovery: Bearer Token { null }
2017-02-08T15:51:09.106646744Z SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
2017-02-08T15:51:09.106794235Z SLF4J: Defaulting to no-operation (NOP) logger implementation
2017-02-08T15:51:09.106899515Z SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
2017-02-08T15:51:09.177266082Z Feb 08, 2017 3:51:09 PM com.hazelcast.spi.discovery.integration.DiscoveryService
2017-02-08T15:51:09.177318318Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Kubernetes Discovery activated resolver: ServiceEndpointResolver
2017-02-08T15:51:09.179664537Z Feb 08, 2017 3:51:09 PM com.hazelcast.instance.Node
2017-02-08T15:51:09.179701224Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Activating Discovery SPI Joiner
2017-02-08T15:51:09.432697612Z Feb 08, 2017 3:51:09 PM com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl
2017-02-08T15:51:09.432748789Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Starting 2 partition threads
2017-02-08T15:51:09.435434911Z Feb 08, 2017 3:51:09 PM com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl
2017-02-08T15:51:09.435452260Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Starting 3 generic threads (1 dedicated for priority tasks)
2017-02-08T15:51:09.444828273Z Feb 08, 2017 3:51:09 PM com.hazelcast.core.LifecycleService
2017-02-08T15:51:09.444873535Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] [10.0.0.18]:5701 is STARTING
2017-02-08T15:51:09.446548445Z Feb 08, 2017 3:51:09 PM com.hazelcast.nio.tcp.nonblocking.NonBlockingIOThreadingModel
2017-02-08T15:51:09.446564564Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] TcpIpConnectionManager configured with Non Blocking IO-threading model: 3 input threads and 3 output threads
2017-02-08T15:51:10.114933294Z Feb 08, 2017 3:51:10 PM com.hazelcast.nio.tcp.InitConnectionTask
2017-02-08T15:51:10.114989406Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Connecting to /10.0.1.13:5701, timeout: 0, bind-any: true
2017-02-08T15:51:10.130930139Z Feb 08, 2017 3:51:10 PM com.hazelcast.nio.tcp.TcpIpConnectionManager
2017-02-08T15:51:10.130981114Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] Established socket connection between /10.0.0.18:41663 and /10.0.1.13:5701
2017-02-08T15:51:16.152590351Z Feb 08, 2017 3:51:16 PM com.hazelcast.internal.cluster.ClusterService
2017-02-08T15:51:16.152666670Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] 
2017-02-08T15:51:16.152677455Z 
2017-02-08T15:51:16.152681767Z Members [2] {
2017-02-08T15:51:16.152685941Z 	Member [10.0.1.13]:5701 - 81ef793b-c1f1-4486-9d59-d7b23f725f7f
2017-02-08T15:51:16.152690336Z 	Member [10.0.0.18]:5701 - a22b972d-a2ae-41c7-aa75-4a7b89167dad this
2017-02-08T15:51:16.152694108Z }
2017-02-08T15:51:16.152697692Z 
2017-02-08T15:51:18.305935099Z Feb 08, 2017 3:51:18 PM com.hazelcast.core.LifecycleService
2017-02-08T15:51:18.305984573Z INFO: [10.0.0.18]:5701 [dev] [3.7.5] [10.0.0.18]:5701 is STARTED
```

That tells me it's working!!! The above configuration is with the REST API lookup, which previously failed due to the SSL error. Would the authors of the project consider moving that dependency up to 1.3.104 or higher? I deliberately didn't go beyond the 1.3.x versions of the fabric8 library to avoid potential errors, but they're up to 1.4.35 in the 1.x range. There's a 2.0.0 out now, but I've no idea what the changes are there. 

Also, as an aside, the java package names and maven artifact ID have changed between what's released on mvn central and what's in the source on Github. is the project becoming part of Hazelcast or is there something else I'm missing there?

**Comment by emrahkocaman** [comment]
Hi @barnyrelph , thanks for the update, this is a nice finding. Let us check the fabric8's never versions.
About the repo ownership, yes hazelcast-kubernetes is becoming an official Hazelcast plugin.

**Comment by barnyrelph** [comment]
@emrahkocaman Thanks for getting back to me, that'd be great to get a version bump at some point. I'll keep an eye on the project here and at Hazelcast. 

Thanks everyone who pitched in on this issue!

**Comment by googlielmo** [comment]
Fixed by #49 
Can you please check @barnyrelph if you have a chance?
